### PR TITLE
#13 Example of simple OAuth2 client

### PR DESCRIPTION
**Pull request by vtepliuk**
Some things left to be improved...

### CODE REVIEW

File: examples/client/app.py [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+
+app = Flask(__name__)
+
+app.secret_key = b'rehgejgbfkhjb'  # FIXME(vtepliuk): Remove in next version.
+
+
+@app.route('/')
+def index():
+    req = Request(method='GET',
+                  url='http://127.0.0.1:5000/oauth/authorize',
```
Comment (mblomdahl) [inline examples/client/app.py]: 3 things that should not be hard coded; pass them in as environment variables instead, e.g.

```bash
export FLASK_APP=app.py
export OAUTH2_BASE_URL=http://127.0.0.1:5000/
export OAUTH2_CLIENT_ID=7368b9a635413b2f153fa032b3de658c
export OAUTH2_CLIENT_SECRET=458c9658bc138ebd752cfc7076bae60e4e27693d3
flask run -h 0.0.0.0 -p 5001 --reload
/.../
```

Then you pick them up in your client application by ...

```python
from os import environ

OAUTH2_BASE_URL = environ.get('OAUTH2_BASE_URL')
OAUTH2_CLIENT_ID = environ.get('OAUTH2_CLIENT_ID')
OAUTH2_CLIENT_SECRET = environ.get('OAUTH2_CLIENT_SECRET')

app = Flask(__name__)
#...
```

And replace (variable) hard coded values with these constants. :)

Same goes for "`app.secret = b'reh..`". Set it outside application code via env variable.

File: examples/client/app.py [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+
+
+@app.route('/login')
+def login():
+    code = request.args.get('code')
+    state = request.args.get('state')
+    client_id = OAUTH2_CLIENT_ID
+    client_secret = OAUTH2_CLIENT_SECRET
+    token_endpoint = OAUTH2_BASE_URL + 'oauth/token'
+    client_redirect_uri = 'http://127.0.0.1:5001/login'
```
Comment (mblomdahl) [inline examples/client/app.py]: Is this reeeallyyyy needed? Please check and get back.

File: ansible/roles/xl_auth/tasks/main.yml [hunk 3]
```diff
@@ -62,7 +62,7 @@
   register: start_xl_auth
   tags: xl_auth
 
-- name: run 'flask create_user -e libris@kb.se -p <my-secret> --is-active --is-admin --force'
+- name: run 'flask create-user -e torimkdevops.se -p <my-secret> --is-active --is-admin --force'
```
Comment (mblomdahl) [inline ansible/roles/xl_auth/tasks/main.yml]: Spelling!

File: examples/client/Dockerfile [hunk 4]
```diff
@@ -0,0 +1,11 @@
+FROM python:2.7
+WORKDIR /client
+EXPOSE 5001
+RUN pip install flask requests
+COPY . .
+ENV FLASK_APP app.py
+ENV APP_SECRET_KEY jflkdsdlskgn
+ENV OAUTH2_BASE_URL http://id.mkdevops.se/
+ENV OAUTH2_CLIENT_SECRET f924d77bfba83c2923433156524db83f69ae9ce13cdad27637ef7fea15cb45019922a104173fcee2066f60c904bbf6616325b9be42317c345f1b95b0bd0b9528ea5c9503890e29720d6368642714c5505e3bcb08804da52c0008d65ba9b1aa3eb7f21631896bc3f4adcae033e31f7e6ec4a0c64a15f26b79c77aa0211cc4a3d5
+ENV OAUTH2_CLIENT_ID 634a4601e29ec2b1e70268c972714d1b
```
Comment (mblomdahl) [inline examples/client/Dockerfile]: Also add a `OAUTH2_CLIENT_REDIRECT_URI`, here and in `app.py` code, currently hard-coded value won't do. :)

File: ansible/roles/oauth2-client/tasks/main.yml [hunk 5]
```diff
@@ -0,0 +1,12 @@
+---
+
+- name: run oauth2-client
+  docker_container:
```
Comment (mblomdahl) [inline ansible/roles/oauth2-client/tasks/main.yml]: Pass in environment variables to Docker container to configure the OAuth2 client here instead. They cannot be part of the Docker *image*, as they are secret and vary depending on target environment, where a container is created based on it.

File: examples/client/app.py (hunk 2 above)
Comment (vtepliuk) [reply examples/client/app.py]: > Is this reeeallyyyy needed? Please check and get back.

Yes, i think, it does not work without redirect_uri.

File: examples/client/app.py [hunk 6]
```diff
@@ -0,0 +1,61 @@
+from flask import Flask, render_template, request, session
+import uuid
+from base64 import b64encode
+from requests import Request
+import requests
+from os import environ
+
+OAUTH2_BASE_URL = environ.get('OAUTH2_BASE_URL')
```
Comment (mblomdahl) [inline examples/client/app.py]: Couple these important variables with assertions, so application will fail to start if these absolutely necessary fields are readable from the environment. E.g.

```python
OAUTH2_BASE_URL = environ.get('OAUTH2_BASE_URL')
if not OAUTH2_BASE_URL:
    raise AssertionError('OAUTH2_BASE_URL missing!')
OAUTH2_CLIENT_ID = environ.get('OAUTH2_CLIENT_ID')
if not OAUTH2_CLIENT_ID:
    # ...
# ...
```

**Pull request merged**